### ISSUE TITLE
avoid hardcoding MPI compiler names in formulae

### ISSUE THREAD

**Issue by davydden** [LLNL/spack#572]
As soon as someone tries to install things with Intel fortran, then `mpicxx`, `mpicc` etc will fail. I saw this in a couple of formulae.

As a user, it is also not clear to me what spack compiler wrappers are and why are they here, i.e. in `openmpi`
```
    def setup_dependent_environment(self, module, spec, dep_spec):
        """For dependencies, make mpicc's use spack wrapper."""
        os.environ['OMPI_CC'] = 'cc'
        os.environ['OMPI_CXX'] = 'c++'
        os.environ['OMPI_FC'] = 'f90'
        os.environ['OMPI_F77'] = 'f77'
```
I would actually expect something like
```
    def setup_dependent_environment(self, module, spec, dep_spec):
        os.environ['MPI_CC'] = 'mpicc'
        os.environ['MPI_CXX'] = 'mpic++'
        os.environ['MPI_FC'] = 'mpif90'
        os.environ['MPI_F77'] = 'mpif77'
```
and then those will be used throughout in each formula where needed.

**Comment by eschnett** [comment]
The former sets environment variables that are used by OpenMPI's `mpicc` etc. to determine what underlying C compiler to call. This is part of the OpenMPI configuration -- it uses environment variables instead of a configuration file. This is different for MPICH, or for other MPI implementations.

If there are MPI implementation that don't use the "standard" MPI wrapper names, then Spack should either set environment variables (such as the one you suggest), or offer compiler wrappers (called `mpicc` etc.) that are guaranteed to work, and that would call the respective real `mpicc` etc.

**Comment by davydden** [comment]
> If there are MPI implementation that don't use the "standard" MPI wrapper names, then Spack should either set environment variables (such as the one you suggest)

I would say environment variables with full path to the underlying compiler would be a straight forward way to go. Say, if it is OpenMPI, then 
```
os.environ['MPI_CC'] = '/path/to/spack/stach/openmpi/bin/mpicc' 
```
is set somewhere inside `OpenMPI` class.
Whereas for Intel MPI it would be something like
```
os.environ['MPI_CC'] = '/externally/provided/intel/mpi/bin/mpiicc' 
```
The latter can only be set by a user in `packages.yaml` to resolve `mpi` dependency as Intel MPI is not free. In real life you would never compile MPI yourself to be used on a cluster, so i guess mainly it boils down to parsing `packages.yaml` and then making available the user-defined settings during compilation of each package.

Same applies to `cxx`,`f77` and `f90`. And `mpiexec` can also be non-standard. Say here on a cluster they have a special wrapper which should be used to run things. That's important to run some quick tests during installation (i.e. `make check`).

> offer compiler wrappers (called mpicc etc.) that are guaranteed to work, and that would call the respective real mpicc etc.

that's also an option. After parsing `packages.yaml`, some wrappers are created which point to the user-defined MPI compilers. My only concern with wrappers is that when you debug a build, it is less obvious what compiler is used. Whereas environment variables got printed to `spack-build.env` file.

**Comment by tgamblin** [comment]
@davydden: Spack uses compiler wrappers to add RPATHs, `-I` flags, and `-L` flags during the build, to make it easier to find the specific dependencies that a spack package is building with.  It also does things like setting `CMAKE_PREFIX_PATH` to point to all the deps in the current package's DAG.   See [compiler interceptors](http://software.llnl.gov/spack/packaging_guide.html#compiler-interceptors) in the docs.

In the case of MPI, this gets complicated because there are two scenarios:

  1. `mpicc` is called from another Spack build.
  2. `mpicc` is called by a user from outside Spack.

The variables you mentioned are for case (1):

        os.environ['OMPI_CC'] = 'cc'
        os.environ['OMPI_CXX'] = 'c++'
        os.environ['OMPI_FC'] = 'f90'
        os.environ['OMPI_F77'] = 'f77'

These *override* the default `CC` that is set within the `mpicc` script.  They set it to Spack's compiler wrappers to ensure that even if a package uses the MPI wrapper scripts *instead* of Spack's compilers, the build will not accidentally pull in the wrong dependencies or fail to add an RPATH.  These are only in the environment *during* a Spack build, and while users can certainly set the outside of Spack, we don't mandate particular values for them.

For case (2), Spack patches the compiler wrappers so that they do the right thing without any environment variables.  See [`filter_compilers`](https://github.com/LLNL/spack/blob/develop/var/spack/repos/builtin/packages/openmpi/package.py#L90) in the `openmpi` package.  It makes sure that the `CC` variable in `mpicc` is set to the path to the *real* compiler, not to the spack wrapper.  This is necessary because if you build `openmpi` with the Spack wrappers, the default value for `CC` is the compiler used to build OpenMPI, which is Spack's wrapper, which will not work outside of Spack.  See also the `mpich` package, which does the same thing.

**Comment by tgamblin** [comment]
@davydden: exporting some common variables like `MPI_CC` to point at the right compilers seems like a good idea to me.  We have discussed in the past exporting module-scope variables for this, or maybe variables defined on the `Package` object in the `spec` DAG, e.g.:

    spec['mpi'].package.mpicc

See #234 and #404 for some discussion on that stuff.  Basically we want to have a contract of some sort for virtual dependencies.  Right now it's done by convention but the convention is inconsistent.  This is up there somewhere on my priority list, but I am not sure how high.  Merging @eschnett's mac fixes, getting module support in (#552), and merging better platform support so that we can have platform-specific `packages.yaml` files and Cray support (#561) are higher.

As for `cc`, `c++`, `f77`, and `f90`, the *properly named* aliases (see #255) to those are already set in `CC`, `CXX`, `F77`, and `FC` within the build environment.  If you want to see exactly what the build environment for a package looks like, try this:

    spack env <spec>

e.g.:

    spack env petsc

**Comment by davydden** [comment]
`spec['mpi'].package.mpicc` would also do, especially if there is a parent class with virtual methods to impose the convention on any other classes which implement MPI dependency, including external one. I think the most difficult part is to make it work with externally provided MPIs, which is always the case for calculations on clusters.

> As for cc, c++, f77, and f90, the properly named aliases (see #255) to those are already set in CC, CXX, F77, and FC within the build environment.

i mean their MPI-counterparts, `mpic++`, `mpif77`, `mpif90`, etc
